vault/expiration.go

```diff
@@ -2340,6 +2408,11 @@ func (le *leaseEntry) nonexpiringToken() bool {
 	return !le.Auth.LeaseEnabled()
 }
 
+// TODO maybe lock RevokeErr once this goes in: https://github.com/hashicorp/vault/pull/11122
+func (le *leaseEntry) isZombie() bool {
```

Comment:
Is the criteria for a lease being a zombie just that it wasn't able to be revoked? I thought it would be if it had failed to be revoked a certain number of times?

Reply:
I think this is the part that is confusing me. If we can check whether a lease is a zombie or not without storing it, why would we save these leases separately instead of just using that dynamic check?

Reply:
in the current code a lease can only become a zombie by failing revocation a certain number of times. in future code, it can also become a zombie if the backing data store tells us it failed to be revoked and is unlikely to be revoked without manual intervention (this is what VAULT-1978 will set us up for).

The goal of zombification is to save cycles attempting to revoke leases that we know can not likely be automatically revoked. A lot of this lease revocation happens shortly after vault startup, and we would like to not attempt revocation if it's likely to fail. Therefore, we must persist some information about the lease (indicating revocation failure) so that on the next vault start we don't attempt to re-revoke it. In bad cases, there can be 100s of 1000s of zombie leases (e.g. if  a lot were stored in a database which became misconfigured or was disconnected from the network).

Does that make sense?

Reply:
the zombie lease cache is actually more for the future API work than anything (zombie status is checked on startup when leases are being loaded). we had discussed removing the cache (therefore hitting storage every time an API request to list zombies is made), but we decided to keep the cache for now